tools/vhost-user-video/vhost-user-video.c

```diff
@@ -0,0 +1,1824 @@
+        handle_get_params_cmd(video, cmd);
+        break;
+    case VIRTIO_VIDEO_CMD_SET_PARAMS_EXT:
+        handle_set_params_cmd(video, cmd);
+        break;
+    case VIRTIO_VIDEO_CMD_QUERY_CONTROL:
+        handle_query_control_cmd(video, cmd);
+        break;
+    case VIRTIO_VIDEO_CMD_GET_CONTROL:
+        handle_get_control_cmd(video, cmd);
+        break;
+    case VIRTIO_VIDEO_CMD_SET_CONTROL:
+        g_error("**** VIRTIO_VIDEO_CMD_SET_CONTROL unimplemented!");
+        break;
+    default:
+        g_error("Unknown VIRTIO_VIDEO command!");
+        break;
+    }
+}
+
+/* for v3 virtio-video spec currently */
+static void
+video_handle_ctrl(VuDev *dev, int qidx)
+{
+    VuVirtq *vq = vu_get_queue(dev, qidx);
+    VuVideo *video = container_of(dev, VuVideo, dev.parent);
+    size_t cmd_len, len;
+
+    struct vu_video_ctrl_command *cmd;
+
+    for (;;) {
+
+        cmd = vu_queue_pop(dev, vq, sizeof(struct vu_video_ctrl_command));
+        if (!cmd) {
+            break;
+        }
+
+        cmd->vq = vq;
+        cmd->error = 0;
+        cmd->finished = false;
+        cmd->dev = dev;
+
+        cmd_len = video_iov_size(cmd->elem.out_sg, cmd->elem.out_num);
+        cmd->cmd_buf = g_malloc0(cmd_len);
+        len = video_iov_to_buf(cmd->elem.out_sg, cmd->elem.out_num,
+                         0, cmd->cmd_buf, cmd_len);
+
+        if (len != cmd_len) {
+            g_warning("%s: command size incorrect %zu vs %zu\n",
+                      __func__, len, cmd_len);
+        }
+
+        /* header is first on every cmd struct */
+        cmd->cmd_hdr = (struct virtio_video_cmd_hdr *) cmd->cmd_buf;
+        /*bswap header */
+        virtio_video_ctrl_hdr_letoh(cmd->cmd_hdr);
+        g_debug("Received %s cmd", vv_cmd_to_string(cmd->cmd_hdr->type));
+        vv_process_cmd(video, cmd);
+    }
+}
+
+static void
+video_queue_set_started(VuDev *dev, int qidx, bool started)
+{
+    VuVirtq *vq = vu_get_queue(dev, qidx);
+
+    g_debug("queue started %d:%d\n", qidx, started);
+
+    switch (qidx) {
+    case 0:
+        vu_set_queue_handler(dev, vq, started ? video_handle_ctrl : NULL);
+        break;
+    default:
+        break;
+    }
+}
+
+/*
+ * video_process_msg: process messages of vhost-user interface
+ *
+ * Any that are not handled here are processed by the libvhost library
+ * itself.
+ */
+static int video_process_msg(VuDev *dev, VhostUserMsg *msg, int *do_reply)
+{
+    VuVideo *r = container_of(dev, VuVideo, dev.parent);
+
+    g_debug("%s: msg %d", __func__, msg->request);
+
+    switch (msg->request) {
+    case VHOST_USER_NONE:
+        g_main_loop_quit(r->loop);
+        return 1;
+    default:
+        return 0;
+    }
+}
+
+static const VuDevIface vuiface = {
+    .set_features = video_set_features,
+    .get_features = video_get_features,
+    .queue_set_started = video_queue_set_started,
+    .process_msg = video_process_msg,
+    .get_config = video_get_config,
+    .set_config = video_set_config,
```

Comment:
This is an empty dummy function, I think you can remove it.

Reply:
Actually, it is required to get the `F_CONFIG` feature bit set when responding to the `GET_PROTOCOL_FEATURES` message. Even if it is a dummy function, we need a set_config function.

https://github.com/aesteve-rh/qemu/blob/virtio_video/subprojects/libvhost-user/libvhost-user.c#L1474

Reply:
I agree. I will look into this in a separate patch in case a change may be preferable :)